### PR TITLE
Add nullable annotations to Http.Abstractions, Http.Features, Connections.Abstractions 

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
I'll undo the MVC bits, but the rest of this should be good to go.

**Comment by pranavkm** [comment]
This is a continuation of https://github.com/dotnet/aspnetcore/pull/18084. Hopefully all of the initial comments listed there should be addressed.

File: src/Http/Headers/src/CacheControlHeaderValue.cs
Comment (JamesNK) [inline src/Http/Headers/src/CacheControlHeaderValue.cs:338]: `NotNullWhen(true)`?

File: src/Http/Headers/src/EntityTagHeaderValue.cs
Comment (JamesNK) [inline src/Http/Headers/src/EntityTagHeaderValue.cs:164]: ParseList.ParseList doesn't allow a null list be EntityTagHeaderValue does. Intended?

**Review by JamesNK — COMMENTED** [review]
Few things I noticed from reviewing public API changes.

File: src/Http/Headers/src/EntityTagHeaderValue.cs
Comment (pranavkm) [reply src/Http/Headers/src/EntityTagHeaderValue.cs:164]: Did you mean TryParseList? I updated those to also allow nulls since the code that this calls is null tolerant.

**Review by JamesNK — APPROVED** [review]
Looks good to me but I'm not an expert in many of these APIs. @Tratcher / @halter73 should review before merge.

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs
Comment (halter73) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:146]: How do we know `_parameters` isn't null here?

File: src/Http/Headers/src/CookieHeaderParser.cs
Comment (halter73) [inline src/Http/Headers/src/CookieHeaderParser.cs:17]: Do we need both `[MaybeNull]` and `?`? Should this also use `[NotNullWhen(true)]`?

File: src/Http/Headers/src/ContentRangeHeaderValue.cs
Comment (halter73) [inline src/Http/Headers/src/ContentRangeHeaderValue.cs:183]: Why is the out param not nullable here?

File: src/Http/Headers/src/BaseHeaderParser.cs
Comment (halter73) [inline src/Http/Headers/src/BaseHeaderParser.cs:18]: Can `[NotNullWhen(true)]` be applied here too?

File: src/Http/Headers/src/EntityTagHeaderValue.cs
Comment (halter73) [inline src/Http/Headers/src/EntityTagHeaderValue.cs:158]: Shouldn't this be the following?

```suggestion
        public static bool TryParse(StringSegment input, [NotNullWhen(true)] out EntityTagHeaderValue? parsedValue)
```

File: src/Http/Headers/src/HeaderUtilities.cs
Comment (halter73) [inline src/Http/Headers/src/HeaderUtilities.cs:42]: It doesn't look like `parameters` is nullable. Why the `!`?

File: src/Http/Headers/src/MediaTypeHeaderValue.cs
Comment (halter73) [inline src/Http/Headers/src/MediaTypeHeaderValue.cs:484]:
```suggestion
        public static bool TryParse(StringSegment input, [NotNullWhen(true)] out MediaTypeHeaderValue? parsedValue)
```

File: src/Http/Headers/src/HttpHeaderParser.cs
Comment (halter73) [inline src/Http/Headers/src/HttpHeaderParser.cs:33]: Why is this marked `[return: MaybeNull]`? It looks like it throws if it's null. If you change this you can remove `!` from some of the callsites.

File: src/Http/Headers/src/CookieHeaderParser.cs
Comment (pranavkm) [reply src/Http/Headers/src/CookieHeaderParser.cs:17]: Removing `MaybeNull` would be the way to go. This can remove a null value while returning true, so `NotNullWhen` doesn't quite work.

File: src/Http/Headers/src/BaseHeaderParser.cs
Comment (pranavkm) [reply src/Http/Headers/src/BaseHeaderParser.cs:18]: Unfortunately not. The `SupportsMultipleValues` pattern makes nullable annotations really awkward to use .

File: src/Http/Headers/src/HttpHeaderParser.cs
Comment (pranavkm) [reply src/Http/Headers/src/HttpHeaderParser.cs:33]: `BaseHeaderParser.TryParseValue` can return null while returning true so this could return a null value in that case. It would really have helped if the Single-value parser and the multi-value parser where separate types so you could declare their behavior in code.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Http/Headers/src/HttpHeaderParser.cs
Comment (halter73) [reply src/Http/Headers/src/HttpHeaderParser.cs:33]: Can we just throw from this method when it would otherwise return null? Do any of the callers actually handle null appropriately? It looks like most callers would blow up with a NullReferenceException.

Comment (pranavkm) [reply src/Http/Headers/src/HttpHeaderParser.cs:33]: Some code paths actually handle this well: https://github.com/dotnet/aspnetcore/blob/master/src/Http/Headers/src/CacheControlHeaderValue.cs#L330-L334. Like I said, you really want to separate single and mullti-valued parsers for this to be expressed correctly, but that's well beyond the scope of this PR.

Comment (halter73) [reply src/Http/Headers/src/HttpHeaderParser.cs:33]: Would it be too big a big change to have the callers that expect the return value could be null to call TryParseValue instead? I get not wanting to make this much bigger though. My browser is really struggling when I go to the files tab of this PR 😆

**Review by halter73 — APPROVED** [review]

File: src/Http/Headers/src/HttpHeaderParser.cs
Comment (pranavkm) [reply src/Http/Headers/src/HttpHeaderParser.cs:33]: Removing all the null-forgiviness from our src files is certainly desirable. But a) I'm too lazy b) fixing internal implementation, which mostly works well today, has marginal utility. Maybe somebody else would want to give this a go.  😸

File: src/Http/Headers/src/MediaTypeHeaderValue.cs
Comment (Tratcher) [inline src/Http/Headers/src/MediaTypeHeaderValue.cs:209]: Wasn't this _parameters to avoid having Parameters allocate if _parameters was null, and GetQuality was null tolerant?

Comment (pranavkm) [reply src/Http/Headers/src/MediaTypeHeaderValue.cs:209]: Nice catch. I missed addressing this from the original PR: https://github.com/dotnet/aspnetcore/pull/18084/files#r362617412

**Pull request merged**
